### DISCUSSION TITLE
Denunciation operation

### DISCUSSION THREAD

**Discussion by sydhds** [massalabs/massa#3113] [Ideas]
## Intro

Currently, nothing prevent a user to produce multiple blocks (for the same Slot) or to endorse a block multiple times.
By allowing nodes to produce a new kind of operations: `Denunciation`; every user will have an incentive of not doing it.
If an invalid Denunciation will just be ignored, a valid Denunciation will slash some `locked rolls`.
Note that this proposal is to prevent against 'rational' user but cannot do anything against 'pathological' user
that just want to disturb or break the blockclique (at any cost).

## Block Denunciation

A "Block Denunciation" is a new kind of "Operation":

* Slot (When what we want to denounce happens)
* PubKey ()
* Proof (So we can verify it really happens)

### Proof

The proof will be:
* (Signature1, Hash1)
* (Signature2, Hash2)

Checking that the proof is valid, consists of the following checks:
* Hash1 != Hash2
* Check Signature1 against (Hash1, Slot)
* Check Signature2 against (Hash2, Slot)

## Endorsement Denunciation

An "Endorsement Denunciation" is a new kind of "Operation":

* Slot (When what we want to denounce happens)
* Proof (So we can verify it really happens)
* Index (Endorsement index: [0, E-1] where E is the endorsement count (constant))

### Proof

The proof will be:
* (Signature1, Hash1)
* (Signature2, Hash2)

Checking that the proof is valid, consists of the following checks:
* Hash1 != Hash2
* Check Signature1 against (Hash1, Slot, Index)
* Check Signature2 against (Hash2, Slot, Index)

## Locked rolls (aka Deferred credit)

When a Denunciation is proven valid, we then need to ensure that the user have sufficient funds. If we cannot puncture on node
balance or staked rolls, we will use 'locked' rolls` (or Deferred credit). Selling a roll will lock it for some time (== 4 cycles). Note that it restricts
the time, a denunciation can be produced (A constant value will be defined for this).

When a denunciation is valid (and if balance is not enough), a roll (staked or locked) will be converted back to coins
and then split like:
* 50% will be burned
* 50% will be added to the Block reward (and thus redistributed to the Block creators, Endorsers & Endors'ed)

## BlockHeader modifications

For the denunciation proof to be trustable, the BlockHeader signature will be modified like:
* Signature(Hash(serialized header) + slot_serialized)

## Endorsement modifications

For the denunciation proof to be trustable, the Endorsement signature will be modified like:
* Signature(Hash(serialized endorsement) + slot_serialized + index)

## Bootstrap modifications

Bootstrap needs to be modified to allow to stream some subtracted locked rolls.

**Comment by damip** [comment]
* Denouncement => Denunciation
* Index for endorsements is from 0 (included) to E-1 (included)
* In the payload, we need to add the public key that created the denounced object
* endorsement proof should contain Index just once (not 3 times)
* stacked => staked

**Reply by sydhds** [reply]
Done

**Comment by sydhds** [comment]
As implementation is almost ready and merged into testnet_22. I'm closing this discussion :-)
